include/networkit/reachability/ReachableNodes.hpp

```diff
@@ -0,0 +1,80 @@
+// networkit-format
+
+#ifndef NETWORKIT_REACHABILITY_REACHABLE_NODES_HPP_
+#define NETWORKIT_REACHABILITY_REACHABLE_NODES_HPP_
+
+#include <vector>
+
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/graph/Graph.hpp>
+
+namespace NetworKit {
+
+/**
+ * @ingroup reachability
+ */
+class ReachableNodes final : public Algorithm {
+public:
+    /**
+     * Determines or estimates the number of reachable nodes from each node in the graph.
+     *
+     * @param G The graph.
+     * @param exact Whether or not to compute the number of reachable nodes exactly. Only used for
+     * directed graphs, on undirected graphs the number of reachable nodes from every node can be
+     * computed in linear time.
+     */
+    ReachableNodes(const Graph &G, bool exact = true);
+
+    /**
+     * Runs the algorithm.
+     */
+    void run() override;
+
+    /**
+     * Returns the number of reachable nodes from the given node @a u. Only available if @a exact is
+     * true.
+     *
+     * @param u A node.
+     * @returns The number of nodes reachable from @a u.
+     */
+    count numberOfReachableNodes(node u) const {
+        assureFinished();
+        if (!exact)
+            throw std::runtime_error("The number of nodes is not computed exactly, run the "
+                                     "algorithm with exact = true.");
+        return reachableLB[u];
+    }
+
+    /**
+     * Returns a lower bound of the number of reachable nodes from the given node @a u.
+     *
+     * @param u A node.
+     * @returns Lower bound of number of nodes reachable from @a u.
+     */
+    count numberOfReachableNodesLB(node u) const { return reachableLB[u]; }
+
+    /**
+     * Returns an upper bound of the number of reachable nodes from the given node @a u.
+     *
+     * @param u A node.
+     * @returns Upper bound of number of nodes reachable from @a u.
+     */
+    count numberOfReachableNodesUB(node u) const {
+        assureFinished();
+        if (G->isDirected())
+            return exact ? reachableLB[u] : reachableUB[u];
+        else
+            return reachableLB[u];
```

Comment:
For the undirected case: Since `reachableLB == reachableUB` (while not requires exact to be `true`), why does `numberOfReachableNodes` returns a runtime_error if exact is false?

Reply:
Only when `exact` is set to `true` then the number of reachable nodes is computed exactly. I think that `numberOfReachableNodes` should not be used if one does not compute the reachable nodes exactly.

Reply:
Hmm ... it is just a bit strange, that this also applies to the undirected case. `exact` is not used in runUndirected() and the documentation also states that `exact` is`Only used for directed graphs, on undirected graphs the number of reachable nodes from every node can be computed in linear time.` I think this may cause confusion.